### PR TITLE
Need to hold a lock when disconnecting or you can trigger JENKINS-28690 style deadlocks

### PR DESCRIPTION
**Pull request by stephenc**
@reviewbybees especially @hadashi and probably @recampbell

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/plugins/mtslavescloud/MansionRetentionStrategy.java [hunk 1]
```diff
@@ -72,32 +78,68 @@ private boolean shouldHaveConnectedByNow(T c) {
...
+                // idleness,
+                // but before we marked it as unavailable for tasks
+                // See JENKINS-23676
+                try {
+                    Thread.sleep(2000);
+                } catch (InterruptedException e) {
+                    LOGGER.log(Level.FINE, "Interrupted while sleeping before removing node {0}", nodeName);
+                    return;
+                }
+
```
Comment (recampbell) [inline src/main/java/com/cloudbees/jenkins/plugins/mtslavescloud/MansionRetentionStrategy.java]: Now that we have queueLock.withLock() it seems to me like this sleep hack is no longer required. Can't you remove this entire try block?

File: src/main/java/com/cloudbees/jenkins/plugins/mtslavescloud/MansionRetentionStrategy.java [hunk 2]
```diff
@@ -72,32 +78,68 @@ private boolean shouldHaveConnectedByNow(T c) {
...
+                            computer.setDisconnectInProgress(false);
+                            synchronized (MansionRetentionStrategy.this) {
+                                disconnectInProgress = false;
+                            }
+                            return;
+                        }
+                        // TODO figure out why this cannot just be computer.getNode().terminate()
+                        try {
+                            for (Executor e : computer.getExecutors()) {
+                                e.interrupt();
```
Comment (recampbell) [inline src/main/java/com/cloudbees/jenkins/plugins/mtslavescloud/MansionRetentionStrategy.java]: Well, I've seen cases where a node is dead, but executors are still "alive" and so couldn't aren't removed properly.

File: src/main/java/com/cloudbees/jenkins/plugins/mtslavescloud/MansionRetentionStrategy.java [hunk 3]
```diff
@@ -72,32 +78,68 @@ private boolean shouldHaveConnectedByNow(T c) {
...
+                            LOGGER.log(FINE, computer.getName() + " is no longer idle, aborting termination.");
+                            // we lost the race -- mark it as back online
+                            computer.setAcceptingTasks(true);
+                            computer.setDisconnectInProgress(false);
+                            synchronized (MansionRetentionStrategy.this) {
+                                disconnectInProgress = false;
+                            }
+                            return;
+                        }
+                        // TODO figure out why this cannot just be computer.getNode().terminate()
```
Comment (recampbell) [inline src/main/java/com/cloudbees/jenkins/plugins/mtslavescloud/MansionRetentionStrategy.java]: Well, I've seen cases where a node is dead, but executors are still "alive" and so couldn't aren't removed properly. This seems to clear that up.

**Comment by recampbell** [comment]
I'm a :bee: as I think this is an improvement and looks generally correct. But it would be better to just get rid of the sleep now that we have proper locking.

File: src/main/java/com/cloudbees/jenkins/plugins/mtslavescloud/MansionRetentionStrategy.java [hunk 4]
```diff
@@ -114,4 +156,55 @@ protected long getIdleMaxTime() {
...
+    
+    private static QueueLock newQueueLock() {
+        try {
+            return new ReflectionPost592QueueLock();
+        } catch (NoSuchMethodException e) {
+            return new Pre592QueueLock();
+        }
+    }
+
+    // TODO replace all this with Queue.withLock once Jenkins 1.592+
```
Comment (recampbell) [inline src/main/java/com/cloudbees/jenkins/plugins/mtslavescloud/MansionRetentionStrategy.java]: Well, I think you could just upgrade the POM to 1.609 and avoid all this. DEV@cloud should be on 1.609.1 by Monday.

Comment (stephenc) [inline src/main/java/com/cloudbees/jenkins/plugins/mtslavescloud/MansionRetentionStrategy.java]: This plugin is available for use on prem also and our official line for this month at least is 1.565.3+ even after this month it only goes to 1.580.3+

File: src/main/java/com/cloudbees/jenkins/plugins/mtslavescloud/MansionRetentionStrategy.java [hunk 5]
```diff
@@ -72,32 +78,68 @@ private boolean shouldHaveConnectedByNow(T c) {
...
+                    Thread.sleep(2000);
+                } catch (InterruptedException e) {
+                    LOGGER.log(Level.FINE, "Interrupted while sleeping before removing node {0}", nodeName);
+                    return;
+                }
+                
+                queueLock.withLock(new Runnable() {
+                    public void run() {
+                        if (!computer.isIdle() && computer.isOnline()) {
+                            LOGGER.log(FINE, computer.getName() + " is no longer idle, aborting termination.");
```
Comment (jtnord) [inline src/main/java/com/cloudbees/jenkins/plugins/mtslavescloud/MansionRetentionStrategy.java]: string concatination in a FINE logging statement?

**Comment by hadashi** [comment]
This looks fixed to me

**Pull request merged**